Add gp_stat_progress_*_summary system views

l-wang opened the pull request:
Created the following system views for greenplum specific progress reporting

- gp_stat_progress_vacuum_summary
- gp_stat_progress_analyze_summary
- gp_stat_progress_cluster_summary
- gp_stat_progress_create_index_summary
- gp_stat_progress_copy_summary
- gp_stat_progress_basebackup_summary

For reviewer:
- It is better to review commit-by-commit
- Things we are looking for feedback
  1. Replicated table are handled differently in most cases
  2. Do we calculate average among just primaries or also include the coordinator (mostly only applied to pg_basebackup)?
  3. For COPY command, ON SEGMENT commands and replicated tables are handled differently
  4. How to define the "pid" column? We report an arbitrary pid from a segment right now. 

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

huansong reviewed (commented):
Great test coverage! Let's add the new tests to the schedule.

For the things requesting feedbacks, my 2cents are:
> Replicated table are handled differently in most cases

Definitely debatable but I think we can just get summation instead of average for replicated tables too. It would be easier to reason about and set simpler expectation for the users:
* pg_* views: one single-segment stats.
* gp_* views: all of the one-segment stats.
* gp_*\_summary views: all-segment stats.

I understand that the reason we thought about getting average for replicated tables is to treat the summary view as if it is a pg_ view running on a PG database. But after all the summary views are for an MPP cluster and adding them up does more good than harm compared to average. For example, imagine `lockers_total=1` in `gp_stat_progress_create_index_summary` for a replicated table, if we get average we could end up having `lockers_total=0`, which doesn't help at all. On the other hand, for things like `blocks_total` I guess it doesn't hurt to use the average value, but summation still feels slightly more sense.

> Do we calculate average among just primaries or also include the coordinator (mostly only applied to pg_basebackup)

I'm not sure about the exact case but I think if the coordinator has meaningful stats we should consider that in the results. 

> For COPY command, ON SEGMENT commands and replicated tables are handled differently

Good change (re `ON SEGMENT`) I think.

> How to define the "pid" column? We report an arbitrary pid from a segment right now.

Could we show the coordinator pid instead? If not I would just leave NULL there.

Chibin commented on src/backend/catalog/system_views_gp_summary.sql:
```diff
@@ -104,19 +104,14 @@ FROM gp_stat_progress_create_index a
-FROM gp_stat_progress_copy a
-         JOIN pg_class c ON a.relid = c.oid
-         LEFT OUTER JOIN gp_distribution_policy d ON c.oid = d.localoid
-GROUP BY a.datid, a.datname, a.relid, a.command, d.policytype, d.numsegments;
```
Good catch
This should be there...

Chibin replied:
We've added it back

Chibin commented on src/test/isolation2/sql/gp_stat_progress_summary_views.sql:
We've added this test to the `isolation2_schedule`

l-wang commented on src/backend/catalog/system_views_gp_summary.sql:
```diff
@@ -103,3 +103,20 @@ FROM gp_stat_progress_create_index a
+    a.datname,
+    a.relid,
+    a.command,
+    case when (select "type" from gp_stat_progress_copy where gp_segment_id = -1) is not null then (select "type" from gp_stat_progress_copy where gp_segment_id = -1) else (select "type" from gp_stat_progress_copy where gp_segment_id = 0) end as "type",
```
Done. 

> Also, when there are two or more COPYs running in parallel, would select "type" from gp_stat_progress_copy where gp_segment_id = ... return more than one rows? If so, would this query still work?

Good call! We updated the query to accommodate same COPY queries running concurrently. We did that by using `sess_id` from `gp_stat_activity` to identify which processes belong to the same session.

l-wang commented:
> > Replicated table are handled differently in most cases
> 
> Definitely debatable but I think we can just get summation instead of average for replicated tables too. It would be easier to reason about and set simpler expectation for the users:
> 
> * pg_* views: one single-segment stats.
> * gp_* views: all of the one-segment stats.
> * gp_*_summary views: all-segment stats.
> 
> I understand that the reason we thought about getting average for replicated tables is to treat the summary view as if it is a pg_ view running on a PG database. But after all the summary views are for an MPP cluster and adding them up does more good than harm compared to average. For example, imagine `lockers_total=1` in `gp_stat_progress_create_index_summary` for a replicated table, if we get average we could end up having `lockers_total=0`, which doesn't help at all. On the other hand, for things like `blocks_total` I guess it doesn't hurt to use the average value, but summation still feels slightly more sense.

@Chibin @huansong and I discussed offline and we decided to stay with the convention of using averaged values for replicated tables because:
- Using average values align with past discussions: https://github.com/greenplum-db/gpdb/issues/14126
- Existing system_views uses average values: re `gp_stat_all_tables_summary`

PS: There is one exception: `COPY <replicated table> TO`, we need to use sum value for this because COPY is only executed on one segment.

l-wang commented:
> > Do we calculate average among just primaries or also include the coordinator (mostly only applied to pg_basebackup)
> 
> I'm not sure about the exact case but I think if the coordinator has meaningful stats we should consider that in the results.

Right. We included the coordinator stats for pg_basebackup. 


> > How to define the "pid" column? We report an arbitrary pid from a segment right now.
> 
> Could we show the coordinator pid instead? If not I would just leave NULL there.

Thanks for the feedback. We now use coordinator's pid for all the summary views.

l-wang commented:
Thanks @huansong for reviewing! We have addressed all the feedbacks and made the following updates:

- Rewrite tests for gp_stat_progress_copy_summary view, we are a lot happier with the tests now.
- Updated gp_stat_progress_copy_summary view definition to handle same COPY query running concurrently. Thanks to @AJR-VMware for teaching us SQL tricks.
- Use coordinator's pid as the `pid` column for the summary views
- Renamed some tests and added them to isolation2_schedule

Please take another look, thanks!

huansong commented on src/backend/catalog/system_views_gp_summary.sql:
```diff
@@ -150,3 +150,15 @@ FROM gp_stat_progress_copy gspc
     LEFT JOIN gp_stat_progress_copy gspc2 ON gspc.pid = gspc2.pid AND gspc2.gp_segment_id = 0
 GROUP BY gspc.datid, gspc.datname, gspc.relid, gspc.command, ac.sess_id, d.policytype, d.numsegments;
+
+CREATE OR REPLACE VIEW gp_stat_progress_basebackup_summary AS
+WITH d as (select count(*) as numsegments from gp_stat_progress_basebackup)
+SELECT
+    coalesce((select pid from gp_stat_progress_basebackup limit 1), NULL) as pid,
+    a.phase,
+    sum(a.backup_total) as backup_total,
+    sum(a.backup_streamed) as backup_streamed,
+    (sum(a.tablespaces_total)/(select numsegments from d)) as tablespaces_total,
+    (sum(a.tablespaces_streamed)/(select numsegments from d)) as tablespaces_streamed
```
nit: can we just do `avg(a.tablespaces_total)` instead of `sum(a.tablespaces_total)/(select numsegments from d)` (and `a.tablespaces_streamed`)?

l-wang replied:
Done.

huansong reviewed (approved):
LGTM!

l-wang commented on src/test/isolation2/expected/gp_stat_progress_summary_views.out:
Good catch! We rename the test and should've removed this file.

l-wang replied:
Done.
